include/Gaffer/ShufflePlug.inl

```diff
@@ -51,69 +51,86 @@ namespace
 
 const IECore::InternedString g_sourceVariable( "source" );
 
+struct ShuffleValues
+{
+	bool enabled;
+	bool deleteSource;
+	std::string sourcePattern;
+};
+
 } // namespace
 
 namespace Gaffer
 {
 
 template<typename T>
-void ShufflesPlug::shuffle( const T &sourceContainer, T &destinationContainer ) const
+T ShufflesPlug::shuffle( const T &sourceContainer ) const
 {
-	std::unordered_set<std::string> newDestinations;
+	T destinationContainer;
 
-	const Context *context = Gaffer::Context::current();
-	Gaffer::Context::EditableScope scope( context );
-
-	for( ShufflePlugIterator it( this ); !it.done(); ++it )
+	size_t i = 0;
+	std::vector<::ShuffleValues> shuffleValues( this->children().size() );	
+	for( auto &plug : ShufflePlug::Range( *this ) )
 	{
-		ShufflePlug *shuffle = it->get();
-		if( !shuffle->enabledPlug()->getValue() )
-		{
-			continue;
-		}
+		shuffleValues[i].enabled = plug->enabledPlug()->getValue();
+		shuffleValues[i].deleteSource = plug->deleteSourcePlug()->getValue();
+		shuffleValues[i].sourcePattern = plug->sourcePlug()->getValue();		
+		++i;
+	}
 
-		const std::string sourcePattern = shuffle->sourcePlug()->getValue();
-		if( sourcePattern.empty() )
-		{
-			continue;
-		}
+	std::unordered_set<std::string> toDelete;
+	std::unordered_set<std::string> newDestinations;
 
-		const std::string destinationPattern = shuffle->destinationPlug()->getValue();
-		if( destinationPattern.empty() )
-		{
-			continue;
-		}
+	Gaffer::Context::EditableScope scope( Gaffer::Context::current() );
 
-		bool deleteSource = shuffle->deleteSourcePlug()->getValue();
+	for( const auto &sourceData : sourceContainer )
+	{
+		scope.set<std::string>( g_sourceVariable, sourceData.first );
 
-		for( const auto &sourceData : sourceContainer )
+		size_t i = 0;
+		for( auto &plug : ShufflePlug::Range( *this ) )
 		{
-			if( !IECore::StringAlgo::match( sourceData.first, sourcePattern ) )
+			auto &shuffle = shuffleValues[i++];
+			if( shuffle.enabled && !shuffle.sourcePattern.empty() )
 			{
-				continue;
+				const std::string destination = scope.context()->substitute( plug->destinationPlug()->getValue() );
```

Comment:
Yeah, its really just for the tests, but I suppose as `ShufflesPlug.shuffle` is public, its also for as-yet-unknown use cases outside of computes. I did have substitutions disabled on the plug originally, I can totally add that back if you like.

Reply:
> I did have substitutions disabled on the plug originally, I can totally add that back if you like.

Yeah, assuming that doesn't cause problems of its own, lets do it.

Reply:
Addressed in 47b31aad5f3e9d61fece3d86a5740f834282ae89

Reply:
Actually, now that destination is no substitutions again, maybe I should pull it out of the loops and add it to the private struct as well.

Reply:
We can't pull it out of the loop because that would break expressions connected into it. It would be possible to optimise for the no-expression case, but we can do that later, possibly as part of introducing the Shuffler we were talking about.